File: src/test/java/com/hiddewieringa/elevator/domain/elevator/model/ElevatorDoorsSagaTest.java
package com.hiddewieringa.elevator.domain.elevator.model;

import com.hiddewieringa.elevator.domain.elevator.ElevatorDoorsClosed;
import com.hiddewieringa.elevator.domain.elevator.ElevatorDoorsOpened;
import com.hiddewieringa.elevator.domain.saga.ElevatorDoorsSaga;
import org.axonframework.test.saga.SagaTestFixture;
import org.junit.Before;
import org.junit.Test;

import java.time.Duration;
import java.util.UUID;

public class ElevatorDoorsSagaTest {

    private SagaTestFixture<ElevatorDoorsSaga> fixture;

    @Before
    public void setUp() {
        fixture = new SagaTestFixture<>(ElevatorDoorsSaga.class);
    }

    @Test
    public void open() {
        ElevatorId elevatorId = new ElevatorId(UUID.randomUUID());

        fixture.givenNoPriorActivity()
                .whenPublishingA(new ElevatorDoorsOpened(elevatorId))
                .expectActiveSagas(1)
                .expectAssociationWith("elevatorId", elevatorId.toString())
                .expectScheduledEvent(Duration.ofSeconds(10), new ElevatorDoorsClosed(elevatorId))
                .expectNoDispatchedCommands();
    }

    @Test
    public void close() {
        ElevatorId elevatorId = new ElevatorId(UUID.randomUUID());

        fixture.givenAPublished(new ElevatorDoorsOpened(elevatorId))
                .whenPublishingA(new ElevatorDoorsClosed(elevatorId))
                .expectActiveSagas(0)
                .expectNoScheduledEvents()
                .expectNoDispatchedCommands();
    }

    @Test
    public void autoClose() {
        ElevatorId elevatorId = new ElevatorId(UUID.randomUUID());

        fixture.givenAPublished(new ElevatorDoorsOpened(elevatorId))
                .whenTimeElapses(Duration.ofSeconds(11))
                .expectNoScheduledEvents()
                .expectActiveSagas(0)
                .expectNoDispatchedCommands();
    }

}

File: src/test/kotlin/com/hiddewieringa/elevator/domain/elevator/model/ElevatorDoorsSagaTest.kt
package com.hiddewieringa.elevator.domain.elevator.model

import com.hiddewieringa.elevator.domain.elevator.ElevatorDoorsClosed
import com.hiddewieringa.elevator.domain.elevator.ElevatorDoorsOpened
import com.hiddewieringa.elevator.domain.saga.ElevatorDoorsSaga
import org.axonframework.test.saga.SagaTestFixture
import org.junit.Before
import org.junit.Test
import java.time.Duration
import java.util.*

class ElevatorDoorsSagaTest {

    private lateinit var fixture: SagaTestFixture<ElevatorDoorsSaga>

    @Before
    fun setUp() {
        fixture = SagaTestFixture(ElevatorDoorsSaga::class.java)
    }

    @Test
    fun open() {
        val elevatorId = ElevatorId(UUID.randomUUID())

        fixture.givenNoPriorActivity()
                .whenPublishingA(ElevatorDoorsOpened(elevatorId))
                .expectActiveSagas(1)
                .expectAssociationWith("elevatorId", elevatorId.toString())
                .expectScheduledEvent(Duration.ofSeconds(10), ElevatorDoorsClosed(elevatorId))
                .expectNoDispatchedCommands()
    }

    @Test
    fun close() {
        val elevatorId = ElevatorId(UUID.randomUUID())

        fixture.givenAPublished(ElevatorDoorsOpened(elevatorId))
                .whenPublishingA(ElevatorDoorsClosed(elevatorId))
                .expectActiveSagas(0)
                .expectNoScheduledEvents()
                .expectNoDispatchedCommands()
    }

    @Test
    fun autoClose() {
        val elevatorId = ElevatorId(UUID.randomUUID())

        fixture.givenAPublished(ElevatorDoorsOpened(elevatorId))
                .whenTimeElapses(Duration.ofSeconds(11))
                .expectNoScheduledEvents()
                .expectActiveSagas(0)
                .expectNoDispatchedCommands()
    }

}
